src/bitboard.cpp

```diff
@@ -138,57 +278,79 @@ namespace {
-        // 's' computed on an empty board. The index must be big enough to contain
+        // 's' computed on an empty board.
+        m.mask = sliding_attack(Pt, s, 0) & ~edges;
+
+        // For 32 bit magics the index must be big enough to contain
         // all the attacks for each possible subset of the mask and so is 2 power
         // the number of 1s of the mask. Hence we deduce the size of the shift to
-        // apply to the 64 or 32 bits word to get the index.
-        Magic& m = magics[s];
-        m.mask  = sliding_attack(pt, s, 0) & ~edges;
-        m.shift = (Is64Bit ? 64 : 32) - popcount(m.mask);
+        // apply to the 64 or 32 bits word to get the index for a non-fixed shift.
+        m.shift32 = 32 - popcount(m.mask);
 
-        // Set the offset for the attacks table of the square. We have individual
-        // table sizes for each square with "Fancy Magic Bitboards".
-        m.attacks = s == SQ_A1 ? table : magics[s - 1].attacks + size;
+        if constexpr (HasPext || !Is64Bit)
+        {
+            // For PEXT or fancy magic indexing, use the starting offset if on the
+            // first square, and use the previous square's end offset as the current
+            // square's starting offset. Rooks are stored in entries 0 through 0x18FFFF,
+            // and bishops are stored in entries 0x190000 through 0x19147F.
+            constexpr int startOffset = Pt == ROOK ? 0 : 0x19000;
+            m.attacks = s == SQ_A1 ? table + startOffset : magics[s - 1].attacks + size;
+        }
+        else
+        {
+            // For magic bitboards indexing we use pre-computed magics values and
+            // offsets. Since we are using the "fixed shift" approach we do not need
+            // to calculate the shift, because the magic product will always resolve
+            // a unique index using a 64-12 bit shift for rooks and 64-9 bit shift
+            // for bishops.
+            m.magic = knownMagic.magic;
+            m.attacks = table + knownMagic.offset;
+        }
 
         // Use Carry-Rippler trick to enumerate all subsets of masks[s] and
-        // store the corresponding sliding attack bitboard in reference[].
-        b = size = 0;
+        // store the corresponding sliding attack bitboard in the attack table.
+        Bitboard occupied = size = 0;
         do {
-            occupancy[size] = b;
-            reference[size] = sliding_attack(pt, s, b);
-
-            if (HasPext)
-                m.attacks[pext(b, m.mask)] = reference[size];
+            occupancy[size] = occupied;
+            reference[size] = sliding_attack(Pt, s, occupied);
+
+            // If using PEXT we don't need magic numbers and can get index directly,
+            // and if using 64 bit then existing magics are pre-computed.
+            if constexpr (HasPext || Is64Bit) {
+                unsigned index = m.index(occupied);
+                m.attacks[index] = reference[size];
```

Comment:
Here is a good opportunity to assert that `attacks[index]` was empty or the same, to verify the validity of the known magics.